src/leap/mx/bounce.py

```diff
@@ -0,0 +1,524 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+"""
+Everything you need to correctly bounce a message!
+
+This is built from the following RFCs:
+
+  * The Multipart/Report Media Type for the Reporting of Mail System
+    Administrative Messages
+    https://tools.ietf.org/html/rfc6522
+
+  * Recommendations for Automatic Responses to Electronic Mail
+    https://tools.ietf.org/html/rfc3834
+
+  * An Extensible Message Format for Delivery Status Notifications
+    https://tools.ietf.org/html/rfc3464
+"""
+
+
+import re
+import socket
+
+from StringIO import StringIO
+from textwrap import wrap
+
+from email.errors import MessageError
+from email.message import Message
+from email.header import decode_header
+from email.utils import formatdate
+from email.utils import parseaddr
+from email.mime.base import MIMEBase
+from email.mime.multipart import MIMEMultipart
+from email.mime.text import MIMEText
+from email.generator import Generator
+from email.generator import NL
+
+from twisted.internet import defer
+from twisted.internet import protocol
+from twisted.internet import reactor
+from twisted.internet.error import ProcessDone
+from twisted.python import log
+
+
+EMAIL_ADDRESS_REGEXP = re.compile("[^@]+@[^@]+\.[^@]+")
+
+
+def _valid_address(address):
+    """
+    Return whether address is a valid email address.
+
+    :param address: An email address candidate.
+    :type address: str
+
+    :return: Wether address is valid.
+    :rtype: bool
+    """
+    return bool(EMAIL_ADDRESS_REGEXP.match(address))
+
+
+@defer.inlineCallbacks
+def bounce_message(orig_msg, reason, bounce_from, bounce_subject):
+    """
+    Bounce a message.
+
+    :param orig_msg: The original message that will be bounced.
+    :type orig_msg: email.message.Message
+    :param reason: The reason for bouncing the message.
+    :type reason: str
+    :param bounce_from: The sender of the bounce message.
+    :type bounce_from: str
+    :param bounce_subject: The subject of the bounce message.
+    :type bounce_subject: str
+    """
+    orig_rpath = orig_msg.get("Return-Path")
+
+    # do not bounce if sender address is invalid
+    _, addr = parseaddr(orig_rpath)
+    if not _valid_address(addr):
+        log.msg(
+            "Will not send a bounce message to an invalid address: %s"
+            % orig_rpath)
+        return
+
+    msg = _build_bounce_message(
+        bounce_from, bounce_subject, orig_msg, reason)
```

Comment:
It could be confusing that `bounce_message` and `_build_bounce_message` have the same parameters but in different order. Is not important at all, but do you mind putting them in the same order?